4단계 - 자동차 경주(우승자)

boy0516 opened the pull request:
4단계 리뷰 부탁드립니다.🙇🏻‍♂️

boy0516 committed:
feat 3단계 - 결과 dto 추가로 인한 코드 수정

- RaceResultDto 추가
- ResultView의 로직 삭제
- 일부 메서드명 변경

boy0516 committed:
test 3단계 - 결과 dto 추가로 인한 테스트 수정

- RaceResultDtoTest 추가
- ResultViewTest 삭제
- RacingTrackTest 결과값 길이 테스트 삭제

boy0516 committed:
style 3단계 - 생성자 개행 스타일 변경

- 생성자 단일 통합
- integer -> int 타입 수정

boy0516 committed:
test 3단계 - 예외 검증 메시지 수정

- RacingTrackTest 메서드, 예외 검증 수정

boy0516 committed:
docs 4단계 - 개발 항목 문서화

boy0516 committed:
feat 4단계 - 자동차 경주 우승자 판별

- 자동차 우승자 판별 로직 추가
- Dto의 UI로직 삭제
- 문자열 입력 유효성 검사 추가

boy0516 committed:
refactor 4단계 - 자동차 경주 우승자 판별

- car status 위치 수정

boy0516 committed:
test 4단계 - 자동차 경주 우승자 판별

- RefereeTest 추가
- startRace 테스트 수
- position_0이하_테스트 수정

boy0516 committed:
refactor 4단계 - 자동차 경주 우승자 판별

- referee 로직 수정

nooose commented on src/test/java/gameboy/gamepack/racinggame/data/vo/PositionTest.java:
```diff
@@ -23,12 +23,12 @@ void add() {
-    @DisplayName("0이하 위치 값 오류 테스트")
-    @ValueSource(ints = {0, -1})
+    @DisplayName("0미만 위치 값 오류 테스트")
+    @ValueSource(ints = {-1})
```
값이 하나인데 ParameterizedTest 테스트를 굳이 사용할 필요는 없을 것 같아요
`@Test`로 해도 되지않을까요?

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -22,4 +36,20 @@ private static int userIntInput() {
+
+    private static String userStringInput() {
+        Scanner scanner = new Scanner(System.in);
+        return scanner.nextLine();
```
잘못 입력하면 재귀 호출하도록 만드셨네요 👍
계속 잘못입력하면 스캐너를 매번 생성할 것 같은데 상수로 올려놓고 재사용하도록 변경하면 좋을 것 같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,19 +1,33 @@
+
     private static void display(String ask) {
         System.out.println(ask);
     }
```
파라미터명이 `ask`네요
display 메서드만 봤을 때 `ask`는 전혀상관 없는 이름인 것 같습니다. 
단순 text, message 같은 이름이 더 적절하지 않을까요?

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,19 +1,33 @@
+    public static List<String> inputRacerName(String ask) {
+        try {
+            display(ask);
+            return List.of(userStringInput().split(",")).stream()
```
`,`도 상수처리하면 좋겠네요

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -22,4 +36,20 @@ private static int userIntInput() {
+        return scanner.nextLine();
+    }
+
+    private static boolean validationRacerName(String name) {
```
보통 boolean 값을 반환할 때
`isXXX` 같은 표현을 써요

지금 같은 경우는 `validationRacerName` 이름으로 봤을 때 
이 메서드가 예외를 터트리는지 boolean 타입을 반환하는지 모를 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -22,4 +36,20 @@ private static int userIntInput() {
+            throw new InvalidRacerNameException("자동차 이름이 " + MAX_NAME_SIZE + "글자를 초과할 수 없습니다.");
+        }
+        return true;
+    }
```
```suggestion
    private static void validateRacerName(String name) {
        if (name.isBlank()) {
            throw new InvalidRacerNameException("자동차 이름은 공백이 될 수 없습니다.");
        }
        if (name.length() > MAX_NAME_SIZE) {
            throw new InvalidRacerNameException("자동차 이름이 " + MAX_NAME_SIZE + "글자를 초과할 수 없습니다.");
        }
    }
```
이럴거면 그냥 void 타입으로 선언하면 될 것 같습니다.
true 값을 리턴하지 않아도 될 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/RacingGamePack.java:
```diff
@@ -3,27 +3,27 @@
-        int driverCount = InputView.inputNumber("자동차 대수는 몇 대 인가요?");
-        racingTrack = new RacingTrack(driverCount);
+    private RacingTrack initializeRacingTrack() {
+        List<String> racerNames = InputView.inputRacerName("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).");
```
해당 메시지를 게임팩이 알 필요는 없을 것 같습니다.
출력을 책임지는 곳에서 관리할 수 있을 것 같아요
`InputView.inputRacerName()`

nooose commented on src/main/java/gameboy/gamepack/racinggame/RacingGamePack.java:
```diff
@@ -3,27 +3,27 @@
-            ResultView.display(raceResult);
-        }
+        RaceResultDto raceResult = racingTrack.startRace(raceCount);
+        ResultView resultView = new ResultView();
+        resultView.display(raceResult);
```
InputView는 static 메소드를 사용했는데 ResultView는 그렇지 않네요
어떤 차이인가요?

boy0516 replied:
static 메소드로 통일했습니다. 그런데 static으로 했을때 장점을 잘 모르겠는데 static 메소드로 할 필요가 있다고 보시나요?

nooose replied:
딱히 큰 이유는 없는데 InputView는 static 메서드를 사용했고 ResultView는 인스턴스화 시켰길래 물어봤어요
상태가 없는 클래스라면 static 메서드를 사용해도 괜찮은 것 같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,36 @@
+import gameboy.gamepack.racinggame.data.vo.*;
 
 public class Car {
-    private Position position;
+
+    private CarStatus status;
```
CarStatus를 단순하게 래핑하고만 있고 Car의 역할이 나타나지 않는 것 같습니다.
Car가 name, position을 필드로 갖고 있는 것과 어떤 차이가 있나요?

boy0516 replied:
RacingTrack 클래스에서 Referee 클래스에게 Car의 Name과 Position을 전달해야하는데 전달을 편하게 하기위해 래핑했습니다.
driver에서 dto로 묶어서 넘겨주는게 나았을까요??
그냥 Car를 통째로 넘기는게 낫겠네요

nooose replied:
~~Referee에게 Car를 그대로 전달하면 되지않나요?~~

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Driver.java:
```diff
@@ -2,21 +2,31 @@
 
+import gameboy.gamepack.racinggame.data.vo.CarStatus;
+
 public class Driver {
```
Driver의 존재도 어떤 이유인가요?
Car만을 사용했을 때와 어떤 차이가 있는지 궁금합니다.

boy0516 replied:
Car의 run은 달리는 행위니까 호출되면 반드시 position값이 상승하고 driver에서 랜덤조건을 통해 run호출 여부를 결정하도록 했습니다. 
현실에서 운전자가 운전을 어떻게 하느냐에 따라 자동차가 달리고 안 달리고가 결정되니까 그렇게 했습니다만 지금 상태에선 큰 의미는 없습니다.

YAGNI 미준수..

nooose replied:
현실을 반영했다고 하셨는데 그러면 Name은 Driver에 넣어야 하는 게 맞지 않나요?

---

객체지향을 할 때 대부분 현실세계를 반영했다고 하는데
반은 맞고 반을 틀린 것 같아요
객체지향에서는 객체를 의인화한다고 표현하는데요

현실세계에서는 수동적이지만 객체자체는 능동적으로 변할 수 있습니다.

현실세계에서는 이름의 길이가 길다 짧다는 사람이 판별하지만 객체지향에서는 Name 클래스 자체가 판별할 수 있습니다.
자동차도 마찬가지입니다. 움직일 수 있냐 없냐는 Driver가 판단할 수 있지만 Car 혼자서도 판단할 수 있다고 생각합니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/CarStatus.java:
```diff
@@ -0,0 +1,22 @@
+package gameboy.gamepack.racinggame.data.vo;
+
+public class CarStatus {
+    private Name name;
+    private Position position;
+
```
name과 position을 묶어서 어떤 장점을 가져왔는지 모르겠습니다. 🤔

boy0516 replied:
depth가 너무 깊어진거같아서 좋아보이진 않네요. 예전에 amims 할때 도메인에서 2depth넘어가면 쓸데없이 복잡해지니까  2depth이하로 만들라고 했던거 같은데 쓸데없이 타고타고 들어가는거 최소화 할까요?

nooose replied:
이유가 있다면 괜찮습니다. 
하지만 지금 상황에서는 CarStatus가 하는 일이 없습니다.
name, position 값을 사용할 때 바이패스만 하는 것 같고 이름과 위치가 묶여서 협력할 일도 없어보입니다.
불필요한 계층만 하나 더 늘어난 느낌이 있습니다.

nooose replied:
address, detailAddress, zipCode 이런 값들은 주소로 볼 수 있기 때문에 
이 3개를 Address라는 클래스로 만드는 게 좋은 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Name.java:
```diff
@@ -0,0 +1,16 @@
+package gameboy.gamepack.racinggame.data.vo;
+
+public class Name {
+    String name;
+
+    public Name() {
+    }
+
+    public Name(String name) {
+        this.name = name;
+    }
+
+    public String getName() {
+        return name;
+    }
+}
```
이것또한 String name 필드를 선언한 것과 어떤 차이인가요?
VO로 만들었다면 VO만의 특별한 로직이 있어야 한다고 생각합니다.

Name 객체를 유지하려면 이름 길이에 대한 유효성 체크 정도는 있어야 할 것 같습니다.

nooose replied:
앞단 InputView에서 검증을 하더라도 
내부적으로 검증을 하는 것이 좋습니다.

로직이 잘못되었거나 실수로 `new Name("")` 같이 빈 이름이 생성될 수 있기 때문입니다.

nooose replied:
관련된 테스트 코드도 포함되면 좋겠네요

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,19 +1,33 @@
         }
     }
 
+    public static List<String> inputRacerName(String ask) {
```
```suggestion
    public static List<Name> inputRacerName(String ask) {
```
Name을 바로 반환해도 되지 않을까요?
Name 생성자에 유효성 로직이 있다면 보장된 Name을 바로 받을 수 있을 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/dto/RaceResultDto.java:
```diff
@@ -0,0 +1,39 @@
+        return new RaceResultDto(raceLogs, winners);
+    }
+
+    @Override
+    public int hashCode() {
+        return this.getClass().hashCode() + raceLogs.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this.hashCode() == obj.hashCode();
+    }
```
2개의 메소드를 오버라이딩 한 이유가 있나요?
사용하지 않는 메소드이고 동등 비교가 제대로 동작할 것 같지 않네요
왜 그럴까요?

boy0516 replied:
hashCode를 가져오는데 raceLogs는 초기화할때 리스트를 파라미터로 받는데 이때 다른 리스트 인스턴스를 받아으면서 초기화하면 각각 다른 hashCode를 가지니까 동등비교가 안될거같네요.(맞나요?)

nooose replied:
RaceLog 인스턴스를 2개 만들어보고 비교해 보면 알 수 있을 것 같아요
equals와 hashcode가 뭔지 검색해 보시면 좋을 것 같아요

---

왜 오버라이드도 했는지 알고싶어요

boy0516 replied:
dto 초기화할때 리스트가 잘 dto에 들어갔는지 테스트코드로 확인할려고했거든요. 그때 equals로 확인하려고 오버라이드 했었는데 초기화 하는코드가 너무 길어지는거같아서 이렇게까지 해서 검증할 내용인가싶어 제외했어요.

nooose replied:
해당 테스트 코드는 성공하셨나요?
제가 봤을 때는 제대로 동작하지 않았을 것 같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Referee {
+
+    private RaceRecorder recorder;
+    private static final ThreadLocal<Integer> winnerPosition = ThreadLocal.withInitial( () -> 0);
```
```suggestion
    private static final ThreadLocal<Integer> winnerPosition = ThreadLocal.withInitial( () -> 0);
    
    private RaceRecorder recorder;
```

클래스 변수는 멤버변수 위에 배치하는 것이 관례입니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Referee {
+
+    private RaceRecorder recorder;
+    private static final ThreadLocal<Integer> winnerPosition = ThreadLocal.withInitial( () -> 0);
```
ThreadLocal을 사용하셨는데 이유가 있나요?

boy0516 replied:
순간 멤버변수에 값을 저장하면 멀티스레드 환경에서 스레드세이프 하지 않다는 생각에 넣었는데 싱글톤이 아니니까 멀티쓰레드여도 필요가 없네요. 빼겠습니다. 😅

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,49 @@
+        return recorder.playback();
+    }
+
+    public Set<String> getWinners() {
```
```
경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).
a,a,a
시도할 회수는 몇 회 인가요?
4
a: ---
a: ---
a: --
a: ---
a: ---
a: --
a: ---
a: ---
a: --
a: ---
a: ---
a: --

a가 최종 우승했습니다.
```
어떤 문제가 있어보이나요?

nooose replied:
우승자 로직이 다소 복잡해 보입니다.
위치가 가장 높은 값을 뽑고 같은 위치를 가진 참가자만 뽑으면 해결됩니다.

boy0516 replied:
동명이 우승하면 다 띄워주는걸로 우선 바꾸긴 했습니다. 

우승자 이름이 같을 경우 둘 다 출력해주면서 구분할 수 있도록 추가 정보를 보여줄지, 입력부터 같은 이름을 받지 않을지 선택해야할거같은데 이런 정책적으로 구분해야할 사항은 어떻게 하나요? 임의로 한쪽으로 결정해서 처리 할까요?

nooose replied:
네 둘중에 어떤 선택을 할지는 지훈님 맘대로 하시면 됩니다

nooose reviewed (changes requested):
모아서 리뷰 남겼다가는 너무 늦게드릴 것 같아서
일단 이정도만 리뷰 남기고 차근차근 나아가기로해요

boy0516 committed:
refactor 4단계 - 자동차 경주 피드백 반영

- Threadloacl 제거
- 우승자 선출 로직 수정
- input 반환 로직 수정
- result 줄바꿈 수정

boy0516 committed:
refactor 4단계 - 자동차 경주 피드백 반영

- 우승자 타입 변경
- 불필요한 ParameterizedTest -> Test

boy0516 committed:
refactor 4단계 - 불필요한 depth 축소

- Driver Class 삭제
- CarStatus 삭제

boy0516 committed:
test 4단계 - depth 축소 연관 테스트 코드 수정

- DriverTest 코드 삭제
- 연관 코드 수정

boy0516 committed:
refactor 4단계 - 피드백 반영

- 승리 판단 car에게 위임
- 매직넘버 사용
- 유효성 체크 추가
- 불필요한 코드 삭제

boy0516 committed:
test 4단계 - 피드백 반영 및 테스트 수정

- 자동차 주행 정지 테스트 분리
- Name 클래스 테스트 추가
- RaceRecorder 클래스 테스트 추가
- 매직넘버 추가
- refactor 영향 코드 수정

nooose commented on src/main/java/gameboy/gamepack/racinggame/RacingGamePack.java:
```diff
@@ -3,27 +3,28 @@
+        int raceCount = InputView.inputRaceCount();
+        RaceResultDto raceResult = racingTrack.startRace(raceCount);
+        ResultView resultView = new ResultView();
+        resultView.display(raceResult);
```
display가 static 메서드인데 인스턴스화 할 필요는 없겠네요
```suggestion
        ResultView.display(raceResult);
```

boy0516 committed:
feat 4단계 - 중복 자동차 명 예외 처리 추가

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,73 @@
+        } catch (InvalidRacerNameException e) {
+            return inputCarsName(e.getMessage());
+        }
+    }
```
validRacerName 메소드는 이제 필요없지 않나요?
`.map(Name::new)` Name 생성자에서 검증을 하기 때문에 이중 체크는 불필요한 것 같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,73 @@
+    private static List<String> split(String text, String separator) {
+        List<String> tokens = List.of(text.split(separator));
+        return tokens.stream().map(String::trim).toList();
+    }
```
좀 더 간결하게 표현할 수 있겠네요
```suggestion
    private static List<String> split(String text) {
        return Stream.of(text.split(CARS_NAME_SEPARATOR)).map(String::trim).toList();
    }
```

separator를 파라미터로 받아서 처리하는 것도 좋아보이는데
메소드가 public일 때 한정입니다.
지금은 private 내부적으로 사용하고 있기 때문에 내부 고정값 상수 `CARS_NAME_SEPARATOR`를 사용하는 게 맞습니다.

nooose replied:
split 메소드가 inputCarsName에서 사용하는데 메서드 거리가 너무 멀어요
이것 말고도 InputView에 있는 메소드 전체를 보시고 배치를 고려해보시면 좋겠습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,81 @@
+    private static final int MAX_NAME_SIZE = 5;
+    private static final String CARS_NAME_SEPARATOR = ",";
+
+    private static final Scanner scanner = new Scanner(System.in);
```
```suggestion
    private static final int MAX_NAME_SIZE = 5;
    private static final String CARS_NAME_SEPARATOR = ",";
    private static final Scanner scanner = new Scanner(System.in);
```

nooose replied:
scanner도 상수인데 `SCANNER`로 표현할 수 있을 것 같습니다.
사실 요즘 ide가 굉장히 좋아서 static이 붙은 변수는 별도로 표시해주긴하지만
일관성있게 대문자로 하는 것도 나쁘지는 않을 것 같네요

아니면 MAX_NAME_SIZE, CARS_NAME_SEPARATOR를 소문자로 통일해도 상관없습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,81 @@
+    }
+
+    public static List<Name> inputCarsName() {
+        return inputCarsName("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).").stream().map(Name::new).toList();
```
모든 사람들이 와이드 모니터를 사용하고 있지 않습니다. 😅
```suggestion
        return inputCarsName("경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).")
        .stream()
        .map(Name::new)
        .toList();
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RacingTrack.java:
```diff
@@ -3,45 +3,50 @@
-        engageDriver(driverCount);
+    private List<Car> cars;
+    private Referee referee;
+    public RacingTrack(List<Car> cars, Referee referee) {
```
```suggestion
    private List<Car> cars;
    private Referee referee;
    
    public RacingTrack(List<Car> cars, Referee referee) {
```

멤버변수와 메소드사이는 개행하는 것이 컨벤션입니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RacingTrack.java:
```diff
@@ -3,45 +3,50 @@
-        }
+    public RacingTrack(List<Car> cars) {
+        this(cars, new Referee());
     }
```
생성자 위치가 바뀐 것 같아요
```java
    public RacingTrack(List<Car> cars) {
        this(cars, new Referee());
    }

    public RacingTrack(List<Car> cars, Referee referee) {
        validationDriverCount(cars);
        this.cars = cars;
        this.referee = referee;
    }
```

항상 코드는 위에서 아래로 내려와야합니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RacingTrack.java:
```diff
@@ -3,45 +3,50 @@
+            cars.add(new Car());
         }
+        return cars;
     }
```
이 생성자는 이제 제거해도 괜찮을 것 같습니다.
참가자는 모두 이름을 가지고 참여하고 있습니다.

테스트를 위해서 존재하고 있다면
테스트 코드에서 `RacingTrack(List<Car> cars)` 생성자를 가지고 테스트 할 수 있습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RacingTrack.java:
```diff
@@ -3,45 +3,50 @@
-        drivers.stream().forEach(Driver::drive);
-        return getDriversPosition();
+    private void validationDriverCount(List<Car> cars) {
+        if (cars.size() == 0) {
```
```suggestion
        if (cars.isEmpty()) {
```
😔

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,35 @@
+    public List<String> getWinnerNames() {
+        RaceLog raceLog = recorder.playbackLastLog();
+        int winnerPosition = getWinnerPosition(raceLog);
+        return raceLog.getCars().stream().filter(car -> car.isWin(winnerPosition)).map(Car::getName).toList();
```
```suggestion
        return raceLog.getCars()
        .stream()
        .filter(car -> car.isWin(winnerPosition))
        .map(Car::getName).toList();
```
체인이 여러개라면 가독성을 위해 끊어주시는 게 좋습니다.

nooose replied:
position을 찾고 해당 포지션과 같은 값을 가지는 것을 찾는게 좋지 않나요?

`filter(car -> car.isSamePosition(winnerPosition))`

`isWin()` 내부 로직을 봤을 때 정말 우승자가 맞는지 애매할 것 같습니다.
그럴바에 해당 위치가 같은 값을 찾는 메서드를 열어두고 우승자를 구하는 Referee에서 메소드를 활용하여
우승자를 구하는 그림이 더 알맞을 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private int getWinnerPosition(RaceLog raceLog) {
+        return raceLog.getCars().stream().mapToInt(Car::getPosition).max().orElse(MIN_POSITION_VALUE);
```
```suggestion
        return raceLog.getCars().stream()
        .mapToInt(Car::getPosition)
        .max()
        .orElse(MIN_POSITION_VALUE);
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,59 @@
+
+    public boolean isWin(int winnerPosition) {
+        return getPosition() >= winnerPosition;
+    }
```
앞에 피드백 드렸듯이
Position이라는 객체를 왜 만드셨나요? 
```suggestion
    public boolean isSamePosition(Position that) {
        return this.position.equals(that);
    }
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Position.java:
```diff
@@ -11,13 +13,13 @@ public Position(int position) {
+    public void addOne() {
         position += 1;
     }
 
```
Position의 equals와 hashcode가 제대로 구현되지 않은 것 같아요
equals는 어떤 상황에서 사용되며 hashcode는 언제 사용될까요?
답변 부탁드립니다.

boy0516 replied:
equals는 객체 인스턴스의 고유 번호를 비교할때 사용되고 hashcode는 고유번호인데 position은 vo로 어떠한 값을 관리하기 위한 클래스니까 객체의 고유번호 비교는 의미 없으니 담고있는 값을 hashcode로 오버라이딩해서 equals로 값을 비교하려는 의도라고 생각했는데 맞나요? 잘 모르겠는데 일단 바꿔서 올립니다.🧐

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,35 @@
+        return recorder.playback();
+    }
+
+    public List<String> getWinnerNames() {
```
지금은 우승자 이름을 반환하고 있는데 
요구사항이 변경되어 우승자의 이름과 나이를 출력해달라고 요청이 온다면 어떻게 될까요?

이러면 반환타입이 `List<String>`에서 `List<T>`로 변경되고
이 메소드를 사용하는 쪽도 변경되고 출력담당쪽도 변경되겠죠
이런 불상사를 막기 위해 객체를 넘겨주는 것이 좋습니다.

그러면 요구사항이 변경되어도 해당 객체에 필드만 추가하고
출력을 담당하는 곳에서는 해당 필드만 갈아끼우거나 추가해도되죠 

```suggestion
    public List<Car> getWinners() {
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Name.java:
```diff
@@ -0,0 +1,31 @@
+
+    String name;
+
+    public Name() {
+    }
```
이 생성자는 불필요해보입니다.
테스트를 위하는 것이라면
객체를 넣어주시는 것이 좋습니다.

`Car("테스트 이름A")`, `Car("테스트 이름B")`
이전 테스트 코드를 유지하기 위해 생성자를 너무 과하게 남겨놓는 경향이 있어보여요
리팩토링을 거치면서 테스트 코드도 같이 깨지는 것이 맞는 것 같아요
깨질 테스트 코드는 빠르게 알아차리고 해결하는 것이 좋습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Name.java:
```diff
@@ -0,0 +1,31 @@
+        this.name = name;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    private void validationName(String name) {
+        if (name.isBlank()) {
```
메소드 배치가..?

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Name.java:
```diff
@@ -0,0 +1,31 @@
+
+    public static final int MAX_NAME_SIZE = 5;
+
+    String name;
```
특별한 경우가 아니라면 private 접근제한자를 붙여주세요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RaceRecorder.java:
```diff
@@ -0,0 +1,37 @@
+    public RaceRecorder() {
+    }
+
+    RaceRecorder(List<RaceLog> logs) {
+        this.logs = logs;
+    }
```
테스트를 위한 생성자같네요
테스트에서 `record()` 메서드를 사용해서 충분히 테스트할 수 있을 것 같아요

아마 Position(1),Position(2) 같은 생성자는 사용가능하다는 리뷰를 드린 것에서 사용하신 것 같은데
엄연히 다른 내용입니다.

돈이라는 Vo가 있을 때 1000원을 생성하냐 2000원을 생성하냐는 자유입니다 말 그대로 값(VO)이니까요
하지만 RaceRecorder는 값이 아니라 그때그때 기록하기 위한 객체로 만드신 것 같아요
심지어 패키지위치도 entity 라는 곳에 있네요
이런 이유로 기록을 잘 되는지는 `record()` 메소드를 사용해서 테스트하면 될 것 같아요

nooose replied:
그리고 entity라는 패키지를 사용하셨는데 이유가 있나요?

boy0516 replied:
음.. entity는 주체적으로 뭔가를 하는 객체라고 생각했고 vo는 값을 담는 객체로 구분해서 패키징했습니다.

nooose replied:
주로 entity는 db 관련해서 나타낼 때 많이 사용합니다. JPA에서도 그렇고요
무의식적으로 도메인 모델이 무조건 db와 매핑되어야 한다는 고정관념이 있을까봐 질문드렸습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RaceRecorder.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private void validationLogs() {
+        if(logs.size() == 0) {
```
```suggestion
        if(logs.isEmpty()) {
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/RacingTrack.java:
```diff
@@ -3,45 +3,50 @@
-        for (int i = 0; i < driverCount; i++) {
-            drivers.add(new Driver());
-        }
+    public RacingTrack(Car... cars) {
+        this(List.of(cars));
```
`public RacingTrack(List<Car> cars)` 생성자를 사용하면 되지 않을까요?
라이브러리 같은 경우에는 편의 생성자로 열어놔도 좋을 것 같은데
지금 상황에서는 굳이 필요한 생성자로는 보이지는 않네요

nooose reviewed (changes requested):
재요청 부탁드립니다.
질문있으면 코멘트, DM, 대면질문 환영합니다.

boy0516 committed:
feat 4단계 - 피드백 반영

- vo 역할 부여
- 불필요 생성자 제거
- 컨벤션 준수
- Car 깊은 복제 추가

boy0516 committed:
refactor 4단계 - 피드백 반영

- hashCode equals 구현
- 중복체크 유지

boy0516 committed:
test 4단계 - 피드백 반영

- 불필요 테스트  제거
- 컨벤션 준수
- Car 깊은 복제 추가

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,74 @@
+    }
+
+    private static List<String> split(String text, String separator) {
+        List<String> carNames = List.of(text.split(separator))
+            .stream()
+            .map(String::trim)
+            .collect(Collectors.toList());
+        return carNames;
```
```suggestion
        return Stream.of(text.split(separator))
                .map(String::trim)
                .toList();
```
같은 코드입니다.

split된 배열을 리스트로 만들고 stream으로 변경하는 방식보다는
Stream으로 바로 넣어주는 게 더 좋을 것 같아요
그리고 `.collect(Collectors.toList())` 으로 최종 연산을 주로 하다보니
자바16부터 `.toList`를 제공합니다.

nooose replied:
```java
    private static List<String> split(String text) {
        return Stream.of(text.split(CARS_NAME_SEPARATOR))
                .map(String::trim)
                .toList();
    }
```

private 메서드인데 separator 파라미터를 받을 필요가 있나요?
재사용도 안 하는 것 같은데 상수 사용하면 될 것 같아요

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/InputView.java:
```diff
@@ -1,25 +1,74 @@
+        return SCANNER.nextLine();
     }
+
+    private static void validDuplicateName(List<String> carNames) {
+        if (carNames.size() > new HashSet<>(carNames).size()) {
+            throw new InvalidRacerNameException("자동차 이름이 중복될 수 없습니다.");
+        }
+    }
```
이 로직은 입력 클래스에서 처리하는 게 아니라
RacingTrack이 가져가는 게 더 적절할 것 같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,43 @@
+            .stream()
+            .map(Car::getPosition)
+            .max(Position::compareTo)
+            .orElse(new Position(MIN_POSITION_VALUE));
```
```suggestion
            .orElse(Position.MIN));
```

nooose replied:
추가로 equals와 hashcode 개념을 정확하게 파악하지 못하신걸로 보입니다.

equals는 객체간 비교를 할 때 주로 사용하는 메서드입니다.
비교하는 로직은 equals를 오버라이드해서 개발자가 커스텀할 수 있습니다.
```java
Person a = Person(age = 10, name = "홍길동")
Person b = Person(age = 20, name = "홍길동")
```
이렇게 있을 때 _나이가 달라도 name 만 같으면 같은 사람이다_ 라고 판별한다고 했을 때는
직접 equals를 구현해줘야겠죠
지훈님이 구현하신 equals는 부작용들이 많이 보여요

그러면 왜 equals와 hashcode를 같이 구현하라고 할까요?
이유는 HashMap이나 HashSet 등에서 내부적으로 object의 hashcode를 호출하고 있기 때문입니다.
equals가 동등하다면 hashcode도 동일하게 나와야겠죠
```java
    @Override
    public int hashCode() {
        return Objects.hash(age, name, value3, ....); // 예시
    }
```
이렇게 필드들을 모두 다 해시체킹할거냐 아니면 몇개일부만 해시체킹할거냐를 정해야합니다.

nooose replied:
인텔리제이에서
`cmd` + `N` 을 누르면 eqauls와 hashcode를 자동생성할 수 있게 도와줍니다.

<img width="255" alt="image" src="https://github.com/g-oop/goop-java-racingcar/assets/36617783/0814b313-9b81-4e2f-a0cd-5a6ac3fefca4">

생성해보시고 지훈님이 작성하신 equals와 hashcode를 한번 비교해보시면 좋겠네요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Position.java:
```diff
@@ -1,37 +1,57 @@
+import java.util.Objects;
 
-    private int position;
+public class Position implements Comparable<Position> {
```
👍

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Position.java:
```diff
@@ -1,37 +1,57 @@
     }
+
+    @Override
+    public int compareTo(Position o) {
```
o 같이 줄여쓰는 변수는 가급적 없었으면 좋겠습니다.

저는 개인적으로 other나 that 같은 변수를 사용합니다.
(저는 this 값과 비교할 때 that을 사용하면 좀 더 좋더라고요 😅)
의미있는 이름을 항상 부여해주세요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,43 @@
+        return recorder.playback();
+    }
+
+    public List<Car> getWinner() {
```
복수니까 복수형이 좋지 않을까요?

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/dto/RaceResultDto.java:
```diff
@@ -0,0 +1,30 @@
+public class RaceResultDto {
+
+    private List<RaceLog> raceLogs;
+    private List<Car> winner;
```
복수형 사용이 좋겠습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,78 @@
+        this.random = random;
+    }
+
+    public static Car copy(Car car) {
+        return new Car(
+            new Name(car.name.getName()),
+            new Position(car.position.value()),
+            car.random
+        );
     }
 
```
`Cor.copy(car)`도 좋은데
`Car 복사차 = car.copy()`
해당 객체의 카피본이 반환된다로 보면 메서드가 더 읽기 자연스럽지 않나요?

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,78 @@
+            position.addOne();
+        }
+    }
+
+    public Position getPosition() {
+        return position;
+    }
+
+    public String getName() {
+        return name.getName();
+    }
+
```
게터 같은 메소드는 run 메소드 아래보다는 생성자 아래에 배치하는 게 좋을 것 같네요

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/ResultView.java:
```diff
@@ -1,19 +1,35 @@
+    }
+
+    private static String getRacePlayback(List<RaceLog> raceLogs) {
+        return raceLogs.stream().map(ResultView::createResult).collect(Collectors.joining("\n"));
```
이 부분은 리뷰가 반영안되어있네요 😔

nooose commented on src/main/java/gameboy/gamepack/racinggame/view/ResultView.java:
```diff
@@ -1,19 +1,35 @@
     }
+
+    private static String getRaceResult(List<Car> winners) {
+        return winners.stream().map(Car::getName).collect(Collectors.joining(", ")) + "가 최종 우승했습니다.";
```
😔

nooose reviewed (changes requested):
이전 질문에 대한 답변도 코멘트에 녹여져있습니다.
읽어보시고 재요청 부탁드려요

boy0516 committed:
refactor 4단계 - 피드백 반영

- equals, hashcode 수정
- 컨벤션 준수

boy0516 committed:
refactor 4단계 - 피드백 반영

- stream 줄바꿈 수정

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Position.java:
```diff
@@ -1,37 +1,56 @@
     }
+
+    @Override
+    public int compareTo(Position that) {
+        if (this.equals(that)) {
+            return 0;
+        }
+        if (this.value > that.value) {
+            return 1;
+        }
+        return -1;
+    }
```
```suggestion
    public int compareTo(Position that) {
        return this.value - that.value;
    }
```
한줄로도 표현이 되겠네요

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/vo/Position.java:
```diff
@@ -1,37 +1,56 @@
     @Override
-    public boolean equals(Object obj) {
-        return this.hashCode() == obj.hashCode();
+    public boolean equals(Object that) {
+        if (that == null || getClass() != that.getClass()) {
+            return false;
+        }
+        return Objects.equals(value, ((Position) that).value);
```
자바16부터 나온 패턴매칭이라는 게 있습니다.
한번 검색해 보시고 학습해보시면 좋겠네요

```suggestion
    @Override
    public boolean equals(Object that) {
        if (this == that) {
            return true;
        }
        if (!(that instanceof Position position)) {
            return false;
        }
        return this.value == position.value;
    }
```

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Referee.java:
```diff
@@ -0,0 +1,43 @@
+            .stream()
+            .map(Car::getPosition)
+            .max(Position::compareTo)
+            .orElse(new Position(MIN_POSITION_VALUE));
```
이 부분은 피드백 반영이 안 되어있네요
Position의 기본값을 매번 생성할 필요가 없습니다. 
default 같은 값들은 캐싱해두고 재사용하는 게 맞습니다.
```suggestion
            .orElse(Position.DEFAULT_ZERO);
```

nooose replied:
orElse와 orElseGet 차이도 공부해 보시면 좋겠습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/dto/RaceResultDto.java:
```diff
@@ -0,0 +1,30 @@
+        return winners;
+    }
+
+    public static RaceResultDto of(List<RaceLog> raceLogs, List<Car> winners) {
+        return new RaceResultDto(raceLogs, winners);
+    }
```
정적팩터리 메소드를 사용하셨는데 
일반 생성자와 어떤 차이가 있나요? 지훈님의 의견 작성부탁드려요

boy0516 replied:
음.. 생성자를 호출하기 이전에 뭔가 의도한 로직을 처리할 때 좋은거같습니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/dto/RaceResultDto.java:
```diff
@@ -0,0 +1,30 @@
+import gameboy.gamepack.racinggame.data.entity.Car;
+import gameboy.gamepack.racinggame.data.vo.RaceLog;
+
+public class RaceResultDto {
+
+    private List<RaceLog> raceLogs;
+    private List<Car> winners;
+
```
이런 값 전달용 dto 같은 경우에는 record 타입을 사용하는 게 좋습니다.
record에 대해서는 잘 아실거라고 생각합니다.

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,81 @@
+        return name.getName();
+    }
+
+    public Car copy() {
```
자동차 자체를 불변으로 만든다면
기록을 위해서 카피할 필요도 사라질 것 같아요
`Car movedCar = car.run()`

nooose commented on src/main/java/gameboy/gamepack/racinggame/data/entity/Car.java:
```diff
@@ -1,23 +1,81 @@
+    private static final int MOVE_RANGE = 10;
+    private static final int MOVE_MIN = 4;
+
+    private Random random;
```
지금 랜덤으로 테스트도 가능한 상태이고 충분히 잘 되고있지만
만약 요구사항이 변경되어
자동차의 움직임이 외부 요인(API 등등)에 의해 판단된다면 어떻게 될까요?
아마 Car 테스트 부터 시작해서 많은 변경이 발생할 것 같습니다.

이런 상황에 유연하게 대처할 수 있는 전략패턴이 있는데요
이 패턴에 대해 검색해 보시고 적용하면 좋을 것 같아요

nooose reviewed (approved):
많이 괴롭혔는데 정말 고생많으셨습니다. 🥲
다음 단계 진행하면서 고민하면 좋을 내용들 포함해서 코멘트 남겼어요
